azure/services/privatedns/spec.go

```diff
@@ -0,0 +1,188 @@
+	AdditionalTags    infrav1.Tags
+}
+
+type RecordSpec struct {
+	Record        infrav1.AddressRecord
+	ZoneName      string
+	ResourceGroup string
+}
+
+// ResourceName returns the name of the private dns zone.
+func (z ZoneSpec) ResourceName() string {
+	return z.Name
+}
+
+// OwnerResourceName is a no-op for private dns zone.
+func (z ZoneSpec) OwnerResourceName() string {
+	return ""
+}
+
+// ResourceGroupName returns the name of the resource group of the private dns zone.
+func (z ZoneSpec) ResourceGroupName() string {
+	return z.ResourceGroup
+}
+
+// Parameters returns the parameters for the private dns zone.
+func (z ZoneSpec) Parameters(existing interface{}) (params interface{}, err error) {
+	_, log, done := tele.StartSpanWithLogger(context.TODO(), "privatedns.ZoneSpec.Parameters")
+	defer done()
+
+	if existing != nil {
+		zone, ok := existing.(privatedns.PrivateZone)
+		if !ok {
+			return nil, errors.Errorf("%T is not a privatedns.PrivateZone", existing)
+		}
+		if !isManaged(zone.Tags, z.ClusterName) {
+			log.V(1).Info("Skipping reconciliation of unmanaged private DNS zone", "private DNS", zone.Name)
+			log.V(1).Info("Tag the DNS manually from azure to manage it with capz."+
+				"Please see https://capz.sigs.k8s.io/topics/custom-dns.html#manage-dns-via-capz-tool", "private DNS", zone.Name)
+			return nil, nil
+		}
+	}
+
+	return privatedns.PrivateZone{
+		Location: to.StringPtr(azure.Global),
+		Tags: converters.TagsToMap(infrav1.Build(infrav1.BuildParams{
+			ClusterName: z.ClusterName,
+			Lifecycle:   infrav1.ResourceLifecycleOwned,
+			Additional:  z.AdditionalTags,
+		})),
+	}, nil
+}
+
+// ResourceName returns the name of the virtual network link.
+func (l LinkSpec) ResourceName() string {
+	return l.Name
+}
+
+// OwnerResourceName returns the zone name of the virtual network link.
+func (l LinkSpec) OwnerResourceName() string {
+	return l.ZoneName
+}
+
+// ResourceGroupName returns the name of the resource group of the virtual network link.
+func (l LinkSpec) ResourceGroupName() string {
+	return l.ResourceGroup
+}
+
+// Parameters returns the parameters for the virtual network link.
+func (l LinkSpec) Parameters(existing interface{}) (params interface{}, err error) {
+	_, log, done := tele.StartSpanWithLogger(context.TODO(), "privatedns.LinkSpec.Parameters")
+	defer done()
+
+	if existing != nil {
+		link, ok := existing.(privatedns.VirtualNetworkLink)
+		if !ok {
+			return nil, errors.Errorf("%T is not a privatedns.VirtualNetworkLink", existing)
+		}
+		if !isManaged(link.Tags, l.ClusterName) {
+			log.V(2).Info("Skipping vnet link reconciliation for unmanaged vnet link", "vnet link",
+				l.Name, "private dns zone", l.ZoneName)
+			return nil, nil
+		}
+	}
+
+	return privatedns.VirtualNetworkLink{
+		VirtualNetworkLinkProperties: &privatedns.VirtualNetworkLinkProperties{
+			VirtualNetwork: &privatedns.SubResource{
+				ID: to.StringPtr(azure.VNetID(l.SubscriptionID, l.VNetResourceGroup, l.VNetName)),
```

Comment:
If you don't need the subscription ID and the VNet name/RG anywhere else, you could just compute the ID in the spec getter instead and remove those fields.

Reply:
This is a [pattern](https://github.com/kubernetes-sigs/cluster-api-provider-azure/blob/7979a9705adabde2085801b04f87ad00e8888e1e/azure/services/loadbalancers/spec.go#L170) that's being used in other services, so didn't want to deviate from it.